### PR TITLE
Allow revocation of direct shares in the Who Has Access view.

### PR DESCRIPTION
**Pull request by dwrensha**
Baby steps towards a more fully-featured Who Has Access view.

This change adds a revoke button for each time the current user appears in the "added by" column. It looks like:

whohasaccess

Clicking the button revokes the share. If the share went through a sharing link, a confirmation dialog is displayed that explains who else might also be affected by the revocation.

### CODE REVIEW

File: shell/client/shell.html [hunk 1]
```diff
@@ -401,8 +401,15 @@ <h4 class="who-has-access">Who has access</h4>
...
             <td> added by
               <ul>
-                {{#each shares}}
-                  <li>{{displayName identityId}}</li>
+                {{#each dedupedShares}}
+                  <li>
+                    {{displayName identityId}}
+                    {{#if isCurrentIdentity identityId}}
+                      <button class="revoke-access" title="revoke" data-recipient="{{../recipient}}">
+                        revoke
```
Comment (kentonv) [inline shell/client/shell.html]: Nitpick: We usually capitalize button titles/labels, I think.

File: shell/shared/grain.js [hunk 2]
```diff
@@ -985,7 +971,116 @@ if (Meteor.isClient) {
...
+      const transitiveShares = instance.transitiveShares.get();
+      const tokensById = instance.downstreamTokensById.get();
+      const recipientShares = _.findWhere(transitiveShares, {recipient: recipient});
+      const currentIdentityId = Accounts.getCurrentIdentityId();
+      const recipientTokens = _.where(recipientShares.allShares, {identityId: currentIdentityId});
+
+      // Two cases:
+      // 1. All of the tokens are direct shares. Easy. Just revoke them.
+      // 2. Some of the links are child tokens. For each, we walk up the chain of parents to the
+      //    root. Then we walk downwards to see whether an other identities would be immediately
```
Comment (kentonv) [inline shell/shared/grain.js]: an -> any

**Comment by dwrensha** [comment]
Comments have been addressed.

**Comment by kentonv** [comment]
Looks like you updated the "title" attributes but not the button body text? Not that the body text is visible anyway, but usually we capitalize that too...

**Comment by dwrensha** [comment]
Ah, OK, got those too now.

**Pull request merged**